### PR TITLE
feat: improve CLI input testability

### PR DESCRIPTION
**Pull request by kenjis**
Needs #8137

**Description**
To improve testability.

- add `InputOutput` and `MockInputOutput` class to CLI

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by MGatner** [comment]
Have to come back to this one.

File: user_guide_src/source/testing/cli.rst [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+*********************
+Using MockInputOutput
+*********************
+
+.. versionadded:: 4.5.0
+
+MockInputOutput
+===============
+
+**MockInputOutput** provides a esay way to write tests for commands that require
```
Comment (michalsn) [inline user_guide_src/source/testing/cli.rst:20]:
```suggestion
**MockInputOutput** provides an easy way to write tests for commands that require
```

File: system/CLI/CLI.php [hunk 2]
```diff
@@ -1137,15 +1117,23 @@ public static function table(array $tbody, array $thead = [])
...
-            // @codeCoverageIgnoreStart
-            echo $string;
+        static::$io->fwrite($handle, $string);
+    }
 
-            return;
-            // @codeCoverageIgnoreEnd
-        }
+    /**
+     * Testing purpose only
```
Comment (MGatner) [inline system/CLI/CLI.php:1124]: Should these receive `@testTag`?

File: system/Test/Mock/MockInputOutput.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+namespace CodeIgniter\Test\Mock;
+
+use CodeIgniter\CLI\InputOutput;
+use CodeIgniter\Test\Filters\CITestStreamFilter;
+use CodeIgniter\Test\PhpStreamWrapper;
+use InvalidArgumentException;
+use LogicException;
+
+class MockInputOutput extends InputOutput
```
Comment (MGatner) [inline system/Test/Mock/MockInputOutput.php:20]:
```suggestion
final class MockInputOutput extends InputOutput
```

File: system/Test/Mock/MockInputOutput.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
+<?php
+
```
Comment (MGatner) [inline system/Test/Mock/MockInputOutput.php:2]: Probably a good idea to start declaring in new files so there are less surprises. 
```suggestion
declare(strict_types=1);

```

File: system/Test/Mock/MockInputOutput.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+use InvalidArgumentException;
+use LogicException;
+
+class MockInputOutput extends InputOutput
+{
+    /**
+     * String to be entered by the user.
+     *
+     * @var array<int, string>
+     * @phpstan-var list<string>
```
Comment (MGatner) [inline system/Test/Mock/MockInputOutput.php:26]: I think we just decided to consolidate these, right?
```suggestion
     * @var list<string>
```

File: user_guide_src/source/changelogs/v4.5.0.rst [hunk 6]
```diff
@@ -98,6 +98,9 @@ Testing
 
 - **DomParser:** The new methods were added ``seeXPath()`` and ``dontSeeXPath()``
   which allows users to work directly with DOMXPath object, using complex expressions.
+- **CLI:** The new ``InputOutput`` class was added and now you can write test
```
Comment (MGatner) [inline user_guide_src/source/changelogs/v4.5.0.rst:101]:
```suggestion
- **CLI:** The new ``InputOutput`` class was added and now you can write tests
```

**Review by MGatner — CHANGES_REQUESTED** [review]

**Comment by MGatner** [comment]
I love! A few comments and suggestions.

File: system/CLI/CLI.php (hunk 2 above)
Comment (kenjis) [reply system/CLI/CLI.php:1124]: Yes, I forgot it.

Comment (kenjis) [reply system/CLI/CLI.php:1124]: Done.

**Comment by kenjis** [comment]
Adding `declare(strict_types=1)` shows a hidden type error.
```
There was 1 error:

1) CodeIgniter\Commands\Database\ShowTableInfoTest::testDbTableShowsDBConfig
TypeError: CodeIgniter\CLI\InputOutput::input(): Return value must be of type string, bool returned

/home/runner/work/CodeIgniter4/CodeIgniter4/system/CLI/InputOutput.php:51
/home/runner/work/CodeIgniter4/CodeIgniter4/system/CLI/CLI.php:259
/home/runner/work/CodeIgniter4/CodeIgniter4/system/CLI/CLI.php:296
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Commands/Database/ShowTableInfo.php:134
/home/runner/work/CodeIgniter4/CodeIgniter4/system/CLI/Commands.php:65
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Common.php:198
/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Commands/Database/ShowTableInfoTest.php:69
```
https://github.com/codeigniter4/CodeIgniter4/actions/runs/6757952382/job/18369383288?pr=7978

**Comment by MGatner** [comment]
Nice strict catch! I'm surprised PHPStan didn't complain. Thanks for the changes, this is looking great.

File: system/CLI/InputOutput.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        echo $prefix;
+
+        $input = fgets(fopen('php://stdin', 'rb'));
+
+        if ($input === false) {
+            $input = '';
+        }
+
+        return $input;
```
Comment (MGatner) [inline system/CLI/InputOutput.php:57]: Are we forbidding the Elvis operator?
```suggestion
        return fgets(fopen('php://stdin', 'rb')) ?: '';
```

**Review by MGatner — APPROVED** [review]
Meant to be "approve" - my comment is not required.

File: system/CLI/InputOutput.php (hunk 7 above)
Comment (kenjis) [reply system/CLI/InputOutput.php:57]: > Disallow short ternary operator (?:) - implies weak comparison, it's recommended to use null coalesce operator (??) or ternary operator with strict condition.
https://github.com/phpstan/phpstan-strict-rules#extra-strict-and-opinionated-rules-for-phpstan

Comment (MGatner) [reply system/CLI/InputOutput.php:57]: Humph.

**Pull request merged**